src/diagonal.works/b6/ingest/compact/encoding.go

```diff
@@ -399,162 +408,288 @@ func (lls *LatLngs) UnmarshalWithoutLength(l int, buffer []byte) int {
 	return i
 }
 
+const ValueTypeBits = 2
+
 type Value interface {
-	Marshal(buffer []byte) int
-	Unmarshal(buffer []byte) int
+	Marshal(TypeAndNamespace, []byte) int
+	Unmarshal(TypeAndNamespace, []byte) int
+	MaxSize() int // Bytes needed for encoding.
+}
+
+func EncodeValueType(t b6.ValueType, v interface{}) interface{} {
+	switch v := v.(type) {
+	case uint32:
+		if e := v << ValueTypeBits; e>>ValueTypeBits != v {
+			panic("Can't encode value type")
+		} else {
+			e |= uint32(t)
+			return e
+		}
+	case uint64:
+		if e := v << ValueTypeBits; e>>ValueTypeBits != v {
+			panic("Can't encode value type")
+		} else {
+			e |= uint64(t)
+			return e
+		}
+	}
+	return nil
+}
+
+func DecodeValue(buffer []byte, t interface{}) (interface{}, int) {
+	switch t.(type) {
+	case uint32:
+		v := binary.LittleEndian.Uint32(buffer[0:])
+		return v >> ValueTypeBits, 4
+	case uint64:
+		v, n := binary.Uvarint(buffer[0:])
+		return v >> ValueTypeBits, n
+	}
+
+	return nil, -1
+}
+
+func fromCompactValue(v Value, s encoding.Strings, nt *NamespaceTable) b6.Value {
+	switch v := v.(type) {
```

Comment:
I'm not sure how big some of these values get, but it may make sense to not decode them into new objects, but rather have an interface that just wraps the encoded value itself for efficiency (which in turn could lead to some interface value changes to b6.Value)

Reply:
// ignore for now / leaving todo